File: leader/leader_g/py/merge_split_files.py
import os
import sys
import re
import argparse


def find_split_files(base_file_path):
    directory = os.path.dirname(base_file_path)
    base_filename = os.path.basename(base_file_path)
    pattern = r"(.+)_part\d{3}(\.[^.]+)$"
    match = re.match(pattern, base_filename)
    if not match:
        print(f"错误: 文件 {base_filename} 不符合拆分文件命名模式")
        return None, None, None
    base_name = match.group(1)
    extension = match.group(2)
    split_pattern = re.compile(
        re.escape(base_name) + r"_part\d{3}" + re.escape(extension) + "$"
    )
    split_files = []
    for filename in os.listdir(directory):
        if split_pattern.match(filename):
            split_files.append(os.path.join(directory, filename))

    def extract_number(file_path):
        filename = os.path.basename(file_path)
        num_match = re.search(r"_part(\d{3})", filename)
        return int(num_match.group(1)) if num_match else 0

    split_files.sort(key=extract_number)
    return split_files, base_name, extension


def merge_split_files(split_files, base_name, extension, output_dir=None):
    if not split_files:
        print("没有找到拆分文件")
        return False
    if output_dir is None:
        output_dir = os.path.dirname(split_files[0])
    merged_filename = f"{base_name}_merged{extension}"
    merged_filepath = os.path.join(output_dir, merged_filename)
    print(f"开始合并 {len(split_files)} 个拆分文件...")
    print(f"输出文件: {merged_filepath}")
    try:
        with open(merged_filepath, "wb") as merged_file:
            for i, split_file in enumerate(split_files, 1):
                print(
                    f"正在处理: {os.path.basename(split_file)} ({i}/{len(split_files)})"
                )
                with open(split_file, "rb") as part:
                    data = part.read()
                    merged_file.write(data)
                print(f"  已添加 {len(data)} 字节")
        merged_size = os.path.getsize(merged_filepath)
        total_split_size = sum(os.path.getsize(f) for f in split_files)
        print(f"\n合并完成!")
        print(
            f"合并后文件大小: {merged_size} 字节 ({merged_size / (1024*1024):.2f} MB)"
        )
        print(
            f"拆分文件总大小: {total_split_size} 字节 ({total_split_size / (1024*1024):.2f} MB)"
        )
        if merged_size == total_split_size:
            print("✓ 文件大小验证通过")
        else:
            print("⚠ 文件大小不匹配，可能存在合并问题")
        return True
    except Exception as e:
        print(f"合并过程中出现错误: {e}")
        if os.path.exists(merged_filepath):
            os.remove(merged_filepath)
        return False


def main():
    parser = argparse.ArgumentParser(description="合并拆分的文件")
    parser.add_argument("file", help="任意一个拆分文件（如：filename_part001.ext）")
    parser.add_argument("-o", "--output", help="输出目录（默认为拆分文件所在目录）")
    if len(sys.argv) > 1:
        args = parser.parse_args()
        split_file_path = args.file
        output_dir = args.output
    else:
        if len(sys.argv) == 1:
            print("请将拆分文件拖放到此脚本上，或通过命令行参数指定文件。")
            input("按回车键退出...")
            return
        split_file_path = sys.argv[1]
        output_dir = None
    if not os.path.isfile(split_file_path):
        print(f"错误: 文件不存在: {split_file_path}")
        input("按回车键退出...")
        return
    print(f"输入文件: {split_file_path}")
    split_files, base_name, extension = find_split_files(split_file_path)
    if split_files is None:
        input("按回车键退出...")
        return
    if not split_files:
        print("未找到相关的拆分文件")
        input("按回车键退出...")
        return
    print(f"找到 {len(split_files)} 个拆分文件:")
    for f in split_files:
        file_size = os.path.getsize(f)
        print(f"  {os.path.basename(f)} ({file_size} 字节)")
    success = merge_split_files(split_files, base_name, extension, output_dir)
    if success:
        print(f"\n合并成功！文件已保存为: {base_name}_merged{extension}")
    else:
        print("\n合并失败！")
    input("按回车键退出...")


if __name__ == "__main__":
    main()

File: leader/leader_g/py/merge-split-files.c
#include <ctype.h>
#include <dirent.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>

#ifdef _WIN32
#include <direct.h>
#include <windows.h>
#define mkdir(path, mode) _mkdir(path)
#define SEPARATOR "\\"
#else
#include <sys/types.h>
#include <unistd.h>
#define SEPARATOR "/"
#endif

#define MAX_PATH_LENGTH 1024
#define MAX_FILES 1000
#define BUFFER_SIZE 8192

typedef struct {
  char path[MAX_PATH_LENGTH];
  int part_number;
} SplitFile;

typedef struct {
  SplitFile files[MAX_FILES];
  int count;
  char base_name[256];
  char extension[64];
} FileCollection;

// 提取文件名中的部分编号
int extract_part_number(const char *filename) {
  const char *part_str = strstr(filename, "_part");
  if (!part_str)
    return -1;

  part_str += 5; // 跳过 "_part"

  // 检查接下来的3个字符是否都是数字
  if (!isdigit(part_str[0]) || !isdigit(part_str[1]) || !isdigit(part_str[2])) {
    return -1;
  }

  return (part_str[0] - '0') * 100 + (part_str[1] - '0') * 10 +
         (part_str[2] - '0');
}

// 比较函数用于排序
int compare_files(const void *a, const void *b) {
  const SplitFile *fileA = (const SplitFile *)a;
  const SplitFile *fileB = (const SplitFile *)b;
  return fileA->part_number - fileB->part_number;
}

// 查找所有拆分文件
int find_split_files(const char *base_file_path, FileCollection *collection) {
  char directory[MAX_PATH_LENGTH];
  char base_filename[256];
  char temp_path[MAX_PATH_LENGTH];

  // 提取目录和文件名
  strcpy(temp_path, base_file_path);

  char *last_slash = strrchr(temp_path, SEPARATOR[0]);
  if (last_slash) {
    *last_slash = '\0';
    strcpy(directory, temp_path);
    strcpy(base_filename, last_slash + 1);
  } else {
    strcpy(directory, ".");
    strcpy(base_filename, temp_path);
  }

  // 解析基础文件名和扩展名
  const char *part_pos = strstr(base_filename, "_part");
  if (!part_pos) {
    printf("错误: 文件 %s 不符合拆分文件命名模式\n", base_filename);
    return 0;
  }

  int base_name_len = part_pos - base_filename;
  strncpy(collection->base_name, base_filename, base_name_len);
  collection->base_name[base_name_len] = '\0';

  const char *ext_pos = strchr(part_pos, '.');
  if (ext_pos) {
    strcpy(collection->extension, ext_pos);
  } else {
    strcpy(collection->extension, "");
  }

  // 查找目录中的文件
  DIR *dirp = opendir(directory);
  if (!dirp) {
    printf("错误: 无法打开目录 %s\n", directory);
    return 0;
  }

  struct dirent *entry;
  collection->count = 0;

  while ((entry = readdir(dirp)) != NULL && collection->count < MAX_FILES) {
    // 检查文件名是否匹配模式
    if (strstr(entry->d_name, collection->base_name) &&
        strstr(entry->d_name, "_part") &&
        strstr(entry->d_name, collection->extension)) {

      int part_num = extract_part_number(entry->d_name);
      if (part_num != -1) {
        // 构建完整路径
        snprintf(collection->files[collection->count].path, MAX_PATH_LENGTH,
                 "%s%s%s", directory, SEPARATOR, entry->d_name);
        collection->files[collection->count].part_number = part_num;
        collection->count++;
      }
    }
  }

  closedir(dirp);

  if (collection->count == 0) {
    printf("未找到相关的拆分文件\n");
    return 0;
  }

  // 按部分编号排序
  qsort(collection->files, collection->count, sizeof(SplitFile), compare_files);

  return 1;
}

// 合并拆分文件
int merge_split_files(const FileCollection *collection,
                      const char *output_dir) {
  if (collection->count == 0) {
    printf("没有找到拆分文件\n");
    return 0;
  }

  char output_path[MAX_PATH_LENGTH];
  if (output_dir && strlen(output_dir) > 0) {
    strcpy(output_path, output_dir);
    // 如果输出目录不存在则创建
    mkdir(output_path, 0755);
  } else {
    // 提取第一个文件的目录
    strcpy(output_path, collection->files[0].path);
    char *last_slash = strrchr(output_path, SEPARATOR[0]);
    if (last_slash) {
      *last_slash = '\0';
    }
  }

  // 构建输出文件路径
  snprintf(output_path + strlen(output_path),
           MAX_PATH_LENGTH - strlen(output_path), "%s%s_merged%s", SEPARATOR,
           collection->base_name, collection->extension);

  printf("开始合并 %d 个拆分文件...\n", collection->count);
  printf("输出文件: %s\n", output_path);

  FILE *output_file = fopen(output_path, "wb");
  if (!output_file) {
    printf("错误: 无法创建输出文件 %s\n", output_path);
    return 0;
  }

  unsigned long total_size = 0;
  int success = 1;

  for (int i = 0; i < collection->count; i++) {
    const char *filename = strrchr(collection->files[i].path, SEPARATOR[0]);
    if (filename)
      filename++;
    else
      filename = collection->files[i].path;

    printf("正在处理: %s (%d/%d)\n", filename, i + 1, collection->count);

    FILE *input_file = fopen(collection->files[i].path, "rb");
    if (!input_file) {
      printf("  错误: 无法打开文件 %s\n", collection->files[i].path);
      success = 0;
      break;
    }

    // 获取文件大小
    fseek(input_file, 0, SEEK_END);
    long file_size = ftell(input_file);
    fseek(input_file, 0, SEEK_SET);

    // 读取并写入文件
    char buffer[BUFFER_SIZE];
    size_t bytes_read;
    unsigned long bytes_written = 0;

    while ((bytes_read = fread(buffer, 1, BUFFER_SIZE, input_file)) > 0) {
      size_t bytes_written_this_time =
          fwrite(buffer, 1, bytes_read, output_file);
      if (bytes_written_this_time != bytes_read) {
        printf("  错误: 写入文件时发生错误\n");
        success = 0;
        break;
      }
      bytes_written += bytes_written_this_time;
    }

    printf("  已添加 %lu 字节\n", bytes_written);
    total_size += bytes_written;

    fclose(input_file);

    if (!success) {
      break;
    }
  }

  fclose(output_file);

  if (!success) {
    remove(output_path); // 删除不完整的输出文件
    return 0;
  }

  // 验证文件大小
  struct stat st;
  if (stat(output_path, &st) == 0) {
    unsigned long merged_size = st.st_size;
    printf("\n合并完成!\n");
    printf("合并后文件大小: %lu 字节 (%.2f MB)\n", merged_size,
           merged_size / (1024.0 * 1024.0));
    printf("拆分文件总大小: %lu 字节 (%.2f MB)\n", total_size,
           total_size / (1024.0 * 1024.0));

    if (merged_size == total_size) {
      printf("✓ 文件大小验证通过\n");
    } else {
      printf("⚠ 文件大小不匹配，可能存在合并问题\n");
    }
  }

  return 1;
}

// 显示用法信息
void print_usage(const char *program_name) {
  printf("用法: %s <拆分文件> [-o 输出目录]\n", program_name);
  printf("示例: %s document_part001.txt\n", program_name);
  printf("       %s document_part001.txt -o C:\\output\n", program_name);
}

int main(int argc, char *argv[]) {
// 设置控制台输出编码为UTF-8
#ifdef _WIN32
  SetConsoleOutputCP(65001);
#endif

  setlocale(LC_ALL, "zh_CN.UTF-8");

  if (argc < 2) {
    printf("请将拆分文件拖放到此程序上，或通过命令行参数指定文件。\n");
    print_usage("merge_split_files.exe");
    printf("按回车键退出...");
    getchar();
    return 1;
  }

  char *input_file = NULL;
  char *output_dir = NULL;

  // 解析命令行参数
  for (int i = 1; i < argc; i++) {
    if (strcmp(argv[i], "-o") == 0 && i + 1 < argc) {
      output_dir = argv[i + 1];
      i++;
    } else if (strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0) {
      print_usage("merge_split_files.exe");
      return 0;
    } else {
      input_file = argv[i];
    }
  }

  if (!input_file) {
    printf("错误: 未指定输入文件\n");
    print_usage("merge_split_files.exe");
    return 1;
  }

  // 检查文件是否存在
  FILE *test_file = fopen(input_file, "rb");
  if (!test_file) {
    printf("错误: 文件不存在: %s\n", input_file);
    printf("按回车键退出...");
    getchar();
    return 1;
  }
  fclose(test_file);

  printf("输入文件: %s\n", input_file);

  FileCollection collection;
  if (!find_split_files(input_file, &collection)) {
    printf("按回车键退出...");
    getchar();
    return 1;
  }

  printf("找到 %d 个拆分文件:\n", collection.count);
  for (int i = 0; i < collection.count; i++) {
    struct stat st;
    if (stat(collection.files[i].path, &st) == 0) {
      const char *filename = strrchr(collection.files[i].path, SEPARATOR[0]);
      if (filename)
        filename++;
      else
        filename = collection.files[i].path;

      printf("  %s (%ld 字节)\n", filename, st.st_size);
    } else {
      const char *filename = strrchr(collection.files[i].path, SEPARATOR[0]);
      if (filename)
        filename++;
      else
        filename = collection.files[i].path;

      printf("  %s (无法获取大小)\n", filename);
    }
  }

  int success = merge_split_files(&collection, output_dir);

  if (success) {
    printf("\n合并成功！\n");
  } else {
    printf("\n合并失败！\n");
  }

  printf("按回车键退出...");
  getchar();

  return success ? 0 : 1;
}
